xarray/core/indexing.py

```diff
@@ -298,17 +299,29 @@ def slice_slice(old_slice: slice, applied_slice: slice, size: int) -> slice:
     return slice(start, stop, step)
 
 
+def slice_slice_by_array(old_slice, array, size):
+    normalized = normalize_slice(old_slice, size)
+    new_indexer = array * normalized.step + normalized.start
+
+    if np.any(new_indexer >= normalized.stop):
+        raise IndexError("indices out of bounds")  # TODO: more helpful error message
+
+    return new_indexer
+
+
 def _index_indexer_1d(old_indexer, applied_indexer, size: int):
-    if isinstance(applied_indexer, slice) and applied_indexer == slice(None):
+    if is_full_slice(applied_indexer):
         # shortcut for the usual case
         return old_indexer
     if isinstance(old_indexer, slice):
         if isinstance(applied_indexer, slice):
             indexer = slice_slice(old_indexer, applied_indexer, size)
         elif isinstance(applied_indexer, integer_types):
             indexer = range(*old_indexer.indices(size))[applied_indexer]  # type: ignore[assignment]
+        elif is_full_slice(old_indexer):
```

Comment:
Does this fix point (3) in https://github.com/pydata/xarray/issues/10311#issuecomment-2917012970

> This is "combined" with the arrayized version of BasicIndexer(slice(None), slice(None), slice(None))